core/src/main/java/org/jboss/sbomer/core/features/sbom/utils/GenericPurlWrapperUtil.java

```diff
@@ -0,0 +1,386 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.sbomer.core.features.sbom.utils;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.cyclonedx.model.component.evidence.Identity;
+import org.cyclonedx.model.component.evidence.Identity.Field;
+import org.cyclonedx.model.component.evidence.Method;
+import org.cyclonedx.model.component.evidence.Method.Technique;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.github.packageurl.MalformedPackageURLException;
+import com.github.packageurl.PackageURL;
+
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class GenericPurlWrapperUtil {
+
+    /*
+     * Help with Jackson Nothing actually uses this yet, the contents of the evidence.identiy.methods.value field are
+     * not well documented and not a great deal of examples See
+     * https://cyclonedx.org/guides/OWASP_CycloneDX-Authoritative-Guide-to-SBOM-en.pdf
+     *
+     * Taking my best guess here for something useful
+     */
+    private static class ValueChanges {
+        private final String from;
+        private final String to;
+
+        public ValueChanges(String from, String to) {
+            this.from = from;
+            this.to = to;
+        }
+
+        public String getFrom() {
+            return from;
+        }
+
+        public String getTo() {
+            return to;
+        }
+    }
+
+    @Getter
+    private final PackageURL packageURL;
+
+    @Getter
+    private final double confidenceScore;
+
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+    /*
+     * Our Regex that covers most the patterns we see on the RCM release area
+     */
+    private static final Pattern[] FILENAME_VERSION_REGEX_STRATEGIES = {
+            Pattern.compile("(?<version>\\d+\\.\\d+\\.\\d+)(?<qualsep>[.-](?<qualifier>Final|[A-Z]+\\d*))?"),
+            Pattern.compile("(?<version>\\d+\\.\\d+)(?<qualsep>[.-](?<qualifier>[A-Z]+\\d*))?"),
+            Pattern.compile("(?<version>\\d+_\\d+)(?<qualsep>[.-](?<qualifier>[A-Z]+\\d*))?") };
```

Comment:
I see there is a reason for using multiple separate regexes, but maybe you also want to pull out the common parts of the pattern into constants to make it a bit easier to read.

Reply:
I'm all for more readable but I need an example of what you mean, I think each pattern is subtly different and there's only three, if we add more in the future with more common elements then I think definitely